### ISSUE TITLE
java.lang.NoSuchFieldError:

### ISSUE THREAD

**Issue by Mr93** [facebook/conceal#143]
why i always get this error ? i copied content of proguard_annotations.pro to my proguard file, my build version is compile 'com.facebook.conceal:conceal:1.0.3@aar'

**Comment by helios175** [comment]
Hello. I was checking if any change on the proguard configuration changed after 1.0.3 and doesn't seem so. I will ask you to confirm a couple of things so we can pinpoint where the problem is.

1) Annotation classes:
`DoNotStrip` class should be in `com.facebook.crypto.proguard.annotations` package. You can add the import to your code and check that package is used.

2) proguard configuration:
send us the proguard file you are using. I understand you're using the one from master branch. Which is much newer version, but should be ok.

It should contain:
`-keep,allowobfuscation @interface com.facebook.crypto.proguard.annotations.DoNotStrip`

3) Finally, can you send us the logcat text with the error?
It's to know for which class it's failing.

Thanks!

**Comment by Mr93** [comment]
1. Yes, DoNotStrip class is in com.facebook.crypto.proguard.annotations package
2. Here is my proguard configuration
[proguard-android.txt](https://github.com/facebook/conceal/files/218708/proguard-android.txt)
3. I can't because i removed "lib" folder in my project (i created this folder before, it contain .SO files). So now i don't get this error "java.lang.NoSuchFieldError". I don't know that problems was solved or not but i have another question. Do you have any documents explain clearly how to use this facebook conceal library because i feel confuse when do follow this instruction http://facebook.github.io/conceal/documentation/. Actually i don't understand this, i think this instruction in this link isn't completely, it miss somethings, so i can't use this library

Many thanks !

**Comment by Mr93** [comment]
Hi, now i can encrypt files (i guess that). Every file i encrypted have size around 30kb, is that ok ? i mean i dont know did i encrypt file in right way or not. And when i decrypt files, the decrypt files always have 0Kb, dont know why, can you explain that ? here is my code:

- Encrypt:
`String path = task.getPath()+"/"+task.getName();
			File test = new File(path);
			Log.d("check_path",path);
			Crypto crypto = new Crypto(new SharedPrefsBackedKeyChain(mContext),new SystemNativeCryptoLibrary());
			if(!crypto.isAvailable()){
				Log.d("check_crypt","error");
				return;
			}
			try {
				OutputStream fileStream = new BufferedOutputStream(new FileOutputStream(test));
				OutputStream outputStream = crypto.getCipherOutputStream(fileStream,new Entity("Password"));
				byte[] buffer = new byte[1024];
				outputStream.write(file2byte(path));
				outputStream.close();
			} catch (FileNotFoundException e) {
				e.printStackTrace();
			} catch (KeyChainException e) {
				e.printStackTrace();
			} catch (CryptoInitializationException e) {
				e.printStackTrace();
			} catch (IOException e) {
				e.printStackTrace();
			}`

- file2byte:
`public static byte[] file2byte(String filePath) {
		byte[] buffer = null;
		try {
			File file = new File(filePath);
			FileInputStream fis = new FileInputStream(file);
			ByteArrayOutputStream bos = new ByteArrayOutputStream();
			byte[] b = new byte[1024];
			int n;
			while ((n = fis.read(b)) != -1) {
				bos.write(b, 0, n);
			}
			fis.close();
			bos.close();
			buffer = bos.toByteArray();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
		return buffer;
	}`

- Decode:
`public void decode(String path,String name){
        File mypath = new File(path+"/"+name);
        Log.d("check_path",mypath.getPath());
        Crypto crypto = new Crypto(new SharedPrefsBackedKeyChain(mContext),
                new SystemNativeCryptoLibrary());

        if (!crypto.isAvailable()) {
            return;
        }


        if (!mypath.exists()) {
            Toast.makeText(mContext, "File is not avaiable", Toast.LENGTH_SHORT).show();
            return;
        }
        try {
            FileInputStream fileStream = new FileInputStream(mypath);
            InputStream inputStream = crypto.getCipherInputStream(fileStream, new Entity("Password"));


            ByteArrayOutputStream out = new ByteArrayOutputStream();

            int read;
            byte[] buffer = new byte[8];

            while ((read = inputStream.read(buffer)) != -1) {
                out.write(buffer, 0, read);
            }

            inputStream.close();
            Log.d("check_byte"," " + out.toByteArray().length);

            //String desPath = Environment.getExternalStorageDirectory()+"/ADA/Music";
            byte2File(out.toByteArray(), path, "test_"+name);

            out.close();

        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(mContext, "Decode error", Toast.LENGTH_SHORT).show();
        }
    }`

- Byte2file:
`public static void byte2File(byte[] buf, String filePath, String fileName) {
        BufferedOutputStream bos = null;
        FileOutputStream fos = null;
        File file = null;
        try {
            File dir = new File(filePath);
            if (!dir.exists() && dir.isDirectory()) {
                dir.mkdirs();
            }
            file = new File(filePath + File.separator + fileName);
            fos = new FileOutputStream(file);
            bos = new BufferedOutputStream(fos);
            bos.write(buf);
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (bos != null) {
                try {
                    bos.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            if (fos != null) {
                try {
                    fos.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }`

Thanks !

**Comment by helios175** [comment]
I'll look closer into that example.

Sizes:
- they should be almost identical
- encrypted files has 34 bytes more than the original (2 bytes prefix + iv length + verification tag length)
So there's clearly a problem.

For what I saw the example looks right, although there are unnecessary steps.
- it loads the whole file in memory and then writes to the cipher stream
- similar thing for decryption

I would rather change file2byte to b file2stream, so it reads file and writes to the output stream passed by param. that output stream being the cipherstream. So the file is never hold complete into memory.

But anyway, that's only for better performance, it should work anyway.
I'll give it a try. I'm curious what could be happening.

The instructions for encrypting are in the old version of the README.md
https://github.com/facebook/conceal/blob/f0d0c6c8d35aa3a5debefa682c63bd9b7cbef2e9/README.md
(the current one is being updated for conceal 1.1).

Basically you're right. You create a SharedPrefsBackedKeyChain. Which is the object that will provide and store the keychain (in this case using the app's shared prefs).

Then you create a crypto with that keychain. Crypto is the object that knows how to encrypt/decrypt.

Then you ask that object to wrap an outputstream (to encode), or to wrap an input stream (to decode). It will use the key provided by the keychain.

**Comment by jkane001** [comment]
I'm having a similar problem. I'm using conceal in a private library, and in the file I reference in  consumerProguardFiles within that library, I have put the lines from the example proguard config file. However, when I import the library (mavenLocal) into my app, it doesn't seem to work. I then tried copying the lines into the app's proguard config, and still no joy.

I then had success by removing those config elements, and replaced them with:

-keep class com.facebook.** { *; }
-keep interface com.facebook.** { *; }
-dontwarn com.facebook.**

That DID work, but obviously, works by not proguarding ANYTHING relating to facebook SDKs.

How can I troubleshoot what's happening (or help you)?

**Comment by leocadiotine** [comment]
@jkane001 Same thing is happening to me (I'm using version 1.1.3). What I've done to avoid removing Proguard to all Facebook libraries (including the ones used by the apps that use my library) was adding:

```
# Facebook conceal
-keep class com.facebook.cypher.** { *; }
-keep class com.facebook.crypto.** { *; }
-keep class com.facebook.android.crypto.** { *; }
```

to both my Proguard rules and my `consumerProguardFiles`.

I will upgrade to v2.x soon and I'll check if this still happen. But I also would like to have a better solution for this case.

**Comment by helios175** [comment]
Conceal 2.0 included some fixes in the proguard rules file, which should fix the issue.
